libdnf/hy-query.c

```diff
@@ -1641,3 +1707,229 @@ hy_query_is_not_empty(HyQuery query)
+        DnfPackage *pkg = dnf_package_new(query->sack, id);
+        guint64 build_time = dnf_package_get_buildtime(pkg);
+        if (build_time <= recent_limit) {
+            MAPCLR(query->result, id);
+        }
+    }
+}
+
+static void
+add_duplicates_to_map(Pool *pool, Map *res, Queue samename, int start_block, int stop_block)
+{
+    Solvable *s_first, *s_second;
+    for (int pos = start_block; pos < stop_block; ++pos) {
+        Id id_first = samename.elements[pos];
+        s_first = pool->solvables + id_first;
+        for (int pos2 = pos + 1; pos2 < stop_block; ++pos2) {
+            Id id_second = samename.elements[pos2];
+            s_second = pool->solvables + id_second;
+            if ((s_first->evr == s_second->evr) && (s_first->arch != s_second->arch)) {
+                continue;
+            }
+            MAPSET(res, id_first);
+            MAPSET(res, id_second);
+        }
+    }
+}
+
+void
+hy_filter_duplicated(HyQuery query)
+{
+    Queue samename;
+    Pool *pool = dnf_sack_get_pool(query->sack);
+
+    hy_query_filter(query, HY_PKG_REPONAME, HY_EQ, HY_SYSTEM_REPO_NAME);
+    hy_query_apply(query);
+
+    hy_query_to_name_ordered_queue(query, &samename);
+
+    Solvable *considered, *highest = 0;
+    int start_block = -1;
+    int i;
+    MAPZERO(query->result);
+    for (i = 0; i < samename.count; ++i) {
+        Id p = samename.elements[i];
+        considered = pool->solvables + p;
+        if (!highest || highest->name != considered->name) {
+            /* start of a new block */
+            if (start_block == -1) {
+                highest = considered;
+                start_block = i;
+                continue;
+            }
+            if (start_block != i - 1) {
+                add_duplicates_to_map(pool, query->result, samename, start_block, i);
+            }
+            highest = considered;
+            start_block = i;
+            continue;
+        }
+    }
+    if ((start_block != -1) && (start_block < i)) {
+        add_duplicates_to_map(pool, query->result, samename, start_block, i);
+    }
+}
+
+int
+hy_filter_unneeded(HyQuery query, DnfSwdb *swdb, const int debug_solver)
+{
+    hy_query_apply(query);
+    GPtrArray *nevras = g_ptr_array_new();
+    Queue id_store;
+    HyGoal goal = hy_goal_create(query->sack);
+    Pool *pool = dnf_sack_get_pool(query->sack);
+
+    HyQuery installed = hy_query_create(query->sack);
+    hy_query_filter(installed, HY_PKG_REPONAME, HY_EQ, HY_SYSTEM_REPO_NAME);
+    hy_query_apply(installed);
+    queue_init(&id_store);
+
+    for (Id id = 1; id < pool->nsolvables; ++id) {
+        if (!MAPTST(installed->result, id))
+            continue;
+        Solvable* s = pool_id2solvable(pool, id);
+        const char* nevra = pool_solvable2str(pool, s);
+        g_ptr_array_add(nevras, (gpointer) g_strdup(nevra));
+        queue_push(&id_store, id);
+    }
+
+    GArray *ingexes = dnf_swdb_select_user_installed(swdb, nevras);
+    g_ptr_array_free(nevras, TRUE);
```

Comment:
Memory leak. The array of pointers to strings is freed and memory allocated for strings is lost.
Simplest solution is to change the line 1878.